src/initialization/InitDistribution.cpp

```diff
@@ -0,0 +1,275 @@
+          pp_dist.get("sigmaT", sigt);
+          pp_dist.get("sigmaPx", sigpx);
+          pp_dist.get("sigmaPy", sigpy);
+          pp_dist.get("sigmaPt", sigpt);
+          pp_dist.query("muxpx", muxpx);
+          pp_dist.query("muypy", muypy);
+          pp_dist.query("mutpt", mutpt);
+
+          impactx::distribution::KVdist KVdist(sigx,sigy,sigt,sigpx,
+                                 sigpy,sigpt,muxpx,muypy,mutpt);
+
+          amrex::Vector<amrex::ParticleReal> x, y, t;
+          amrex::Vector<amrex::ParticleReal> px, py, pt;
+          amrex::RandomEngine rng;
+          amrex::ParticleReal ix, iy, it, ipx, ipy, ipt;
+
+          if (amrex::ParallelDescriptor::IOProcessor()) {
+              x.reserve(npart);
+              y.reserve(npart);
+              t.reserve(npart);
+              px.reserve(npart);
+              py.reserve(npart);
+              pt.reserve(npart);
+
+              for(amrex::Long i = 0; i < npart; ++i) {
+                  KVdist(ix, iy, it, ipx, ipy, ipt, rng);
+                  x.push_back(ix);
+                  y.push_back(iy);
+                  t.push_back(it);
+                  px.push_back(ipx);
+                  py.push_back(ipy);
+                  pt.push_back(ipt);
+              }
+          }
+
+          int const lev = 0;
+          m_particle_container->AddNParticles(lev, x, y, t, px, py, pt,
+                                              qm, bunch_charge);
+
+        } else {
+            amrex::Abort("Unknown distribution: " + distribution_type);
+        }
+
+
+        // reference particle
+        amrex::ParticleReal massE;  // MeV
+        if (particle_type == "electron") {
+            massE = 0.510998950;
+        } else if (particle_type == "proton") {
+            massE = 938.27208816;
+        } else {
+            massE = 0.510998950;  // default to electron
+        }
+        RefPart refPart;
+        refPart.x = 0.0;
+        refPart.y = 0.0;
+        refPart.t = 0.0;
+        refPart.px = 0.0;
+        refPart.py = 0.0;
+        refPart.pt = -energy/massE - 1.0_prt;
+        m_particle_container->SetRefParticle(refPart);
+
+        // print information on the initialized beam
+        amrex::Print() << "Beam kinetic energy (MeV): " << energy << std::endl;
+        amrex::Print() << "Bunch charge (C): " << bunch_charge << std::endl;
+        amrex::Print() << "Particle type: " << particle_type << std::endl;
+        amrex::Print() << "Number of particles: " << npart << std::endl;
+        amrex::Print() << "Beam distribution type: " << distribution_type << std::endl;
+
+        if (unit_type == "static") {
+            amrex::Print() << "Static units" << std::endl;
+        } else if (unit_type == "dynamic") {
+            amrex::Print() << "Dynamic units" << std::endl;
+        } else {
+            amrex::Abort("Unknown units (static/dynamic): " + unit_type);
+        }
+
+        amrex::Print() << "Initialized beam distribution parameters" << std::endl;
+    }
```

Comment:
In Impact-Z, the same parameter input set is used for each element (although certain parameters may be interpreted differently in a few cases).  If we follow the same strategy here, much of the code duplication above can be reduced.  However, allowing different types to have different inputs does add some flexibility.  Maybe this could be useful at some future point.  Is it worth it?

Reply:
I would probably just add a helper function for the block
```C++
          if (amrex::ParallelDescriptor::IOProcessor()) {
              x.reserve(npart);
              y.reserve(npart);
              t.reserve(npart);
              px.reserve(npart);
              py.reserve(npart);
              pt.reserve(npart);

              for(amrex::Long i = 0; i < npart; ++i) {
                  KVdist(ix, iy, it, ipx, ipy, ipt, rng);
                  x.push_back(ix);
                  y.push_back(iy);
                  t.push_back(it);
                  px.push_back(ipx);
                  py.push_back(ipy);
                  pt.push_back(ipt);
              }
          }
```

for now. All we need is a template for the exact function we call.

The rest is then specific to the function.

Reply:
->  #108 :)